codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -0,0 +1,457 @@
+                        *codegenScope
+                    ) {
+                        objectKeyLoop {
+                            rustTemplate(
+                                """
+                                if variant.is_some() {
+                                    return Err(#{Error}::custom("encountered mixed variants in union"));
+                                }
+                                """,
+                                *codegenScope
+                            )
+                            withBlock("variant = match key.as_escaped_str() {", "};") {
+                                for (member in shape.members()) {
+                                    val variantName = member.memberName.toPascalCase()
+                                    rustBlock("${member.wireName().dq()} =>") {
+                                        withBlock("Some(#T::$variantName(", "))", symbol) {
+                                            deserializeMember(member, false)
+                                        }
+                                    }
+                                }
+                                // TODO: Handle unrecognized union variants
+                                rust("_ => None")
+                            }
+                        }
+                    }
+                    rustTemplate(
+                        "_ => return Err(#{Error}::custom(\"expected start object or null\"))",
+                        *codegenScope
+                    )
+                }
+                rust("return Ok(variant);")
+            }
+        }
+        rust("#T(tokens)?", nestedParser)
+    }
+
+    private fun RustWriter.objectKeyLoop(inner: RustWriter.() -> Unit) {
+        rustBlock("loop") {
+            rustBlock("match tokens.next().transpose()?") {
+                rustBlockTemplate(
+                    """
+                        Some(#{Token}::EndObject { .. }) => break,
+                        Some(#{Token}::ObjectKey { key, .. }) =>
+                        """,
+                    *codegenScope
+                ) {
+                    inner()
+                }
+                rustTemplate(
+                    "_ => return Err(#{Error}::custom(\"expected object key or end object\"))",
+                    *codegenScope
+                )
+            }
+        }
+    }
+
+    private fun RustWriter.startArrayOrNull(inner: RustWriter.() -> Unit) = startOrNull("array", inner)
+    private fun RustWriter.startObjectOrNull(inner: RustWriter.() -> Unit) = startOrNull("object", inner)
+    private fun RustWriter.startOrNull(objectOrArray: String, inner: RustWriter.() -> Unit) {
+        rustBlockTemplate("match tokens.next().transpose()?", *codegenScope) {
+            rustBlockTemplate(
+                """
+                    Some(#{Token}::ValueNull { .. }) => return Ok(None),
+                    Some(#{Token}::Start${StringUtils.capitalize(objectOrArray)} { .. }) =>
```

Comment:
if it's restricted to this private function maybe just pass `"Array"` and `"Object"`?